### ISSUE TITLE
Better validation message for unexpected object type for $ref properties

### ISSUE THREAD

**Issue by tfesenko** [RepreZen/KaiZen-OpenAPI-Editor#354]
Extracted from #323 "Validation of $ref property values should show error on unexpected object type"
The current error message just says "Invalid object reference, the referenced object is not of expected type."  It does not provide information about the expected type:

<img width="768" alt="screen shot 2017-06-08 at 5 22 17 pm" src="https://user-images.githubusercontent.com/644582/26952443-7bd8eb8a-4c73-11e7-8794-2a20a820e009.png">

We should see which information we can get from the JSON Schema and provide it in the message

**Comment by ghillairet** [comment]
One possible solution is to give the pointer of the current type, and the pointer of the expected type, like this:

![screen shot 2017-06-09 at 16 53 02](https://user-images.githubusercontent.com/148045/26981038-444792e0-4d34-11e7-8283-2e34fce1dec9.png)

Test model:

```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Swagger Petstore
paths:
  /:
    get:
      responses:
        200:
          $ref: "#/components/schemas/Pet"             
components:
  responses:
    Ok:
      description: ok
  schemas:
    Pet:
      properties:
        id:
          type: number
```

**Comment by tedepstein** [comment]
This is an example of where "title" could come in very handy.

**Comment by ghillairet** [comment]
Unfortunately `title` is not available everywhere in the schema, for example `responseOrReference` is defined like this:

```json
"responseOrReference": {
      "oneOf": [
        {
          "$ref": "#/definitions/response"
        },
        {
          "$ref": "#/definitions/reference"
        }
      ]
    },
```

And response is defined like this:

```json
"response": {
      "type": "object",
      "description": "Describes a single response from an API Operation, including design-time, static  `links` to operations based on the response.",
      "required": [
        "description"
      ],
      "additionalProperties": false,
      "patternProperties": {
        "^x-": {
          "$ref": "#/definitions/specificationExtension"
        }
      },
      "properties": {
        "description": {
          "type": "string"
        },
        "headers": {
          "$ref": "#/definitions/headers"
        },
        "content": {
          "$ref": "#/definitions/content"
        },
        "links": {
          "$ref": "#/definitions/links"
        }
      }
    }
```

I could change the message to include only the last path of the pointer, so it would be `It should be of type responseOrReference but it is of type schemaOrReference`.

**Comment by tfesenko** [comment]
Somewhat similar code for the `oneOf` validation cases:
`com.reprezen.swagedit.core.validation.MultipleSwaggerErrorBuilder.getHumanFriendlyText(String)`

**Comment by tedepstein** [comment]
> Unfortunately title is not available everywhere in the schema.

Right, we would have to add it.  It would be an optimization.  Using the last segment of the JSON pointer to the definition would be the fallback strategy for schemas that don't have titles.  And honestly, it is probably good enough for most cases. But we could consider adding title later on, and making the algo use title wherever available.

**Comment by tedepstein** [comment]
@ghillairet [wrote above](https://github.com/RepreZen/KaiZen-OpenAPI-Editor/issues/354#issuecomment-307411314): 

> One possible solution is to give the pointer of the current type, and the pointer of the expected type, like this...

The problem is that we won't always be able to determine exactly one "type" (i.e. JSON Schema) of the value that the reference resolves to.  There are lots of edge cases:
 - The reference could resolve to a primitive value or array, instead of an object. 
 - The reference could resolve to an object that conforms to more than one of our known schemas. 
 - The object might not be valid against any of our known schemas. 

It would be better to show the schema validation error message, using the same presentation logic that we use for validation errors on inline objects.  

For example, an inline `Response Object` with an unrecognized property shows a validation message like this:

![image](https://user-images.githubusercontent.com/692813/39583042-39cb0bea-4ebd-11e8-9104-69bd44faf80c.png)

A reference to that invalid Response Object should show the generic message, "Invalid object reference, the referenced object is not of expected type," followed by the same schema validation message that is shown on the invalid Response Object, above. 

![image](https://user-images.githubusercontent.com/692813/39583209-9a4d6634-4ebd-11e8-8a4f-dcd3774ffabb.png)

It might also be helpful to specify the name of the expected schema, or (a bit better) the name of the corresponding object name in the OpenAPI spec, if there is exactly one such object.

So for the above example, the final error message could be:

> Invalid object reference. A Response Object was expected, and the referenced object does not conform to the expected type:
> object has properties "errorCode" which are not allowed
